### DISCUSSION TITLE
Optional params in layoutData

### DISCUSSION THREAD

**Discussion by Ahrengot** [livewire/livewire#2820] [1. Help]
Hey everybody!

I'm using full page components more and more, but I'm running into an issue where I have to keep sending null values via layoutData and I'm hoping there's some kind of way around this.

My main layout file has a bunch of parameters, all of which are optional. The constructor in AppLayout.php looks like this:

```PHP
public function __construct($title = "", $headline = "", $fullBleed = false, $breadcrumbLinks = [], $sidebar = null, $tabBarActions = null)
{
    $this->title = $title;
    $this->headline = $headline;
    $this->fullBleed = $fullBleed;
    $this->breadcrumbLinks = $breadcrumbLinks;
    $this->sidebar = $sidebar;
    $this->tabBarActions = $tabBarActions;
}
```
In my blade templates I can use the layout without passing any of the params e.g.:

```HTML
<x-app-layout>
    <p>Page content here</p>
</x-app-layout>
```

Or I can pass any or all of the accepted params

```HTML
<x-app-layout title="Some page title" fullBleed>
    <x-slot name="sidebar">
        My sidebar
    </x-slot>

    <p>Page content here</p>
</x-app-layout>
```

However, with Livewire and `layoutData()`, I have to explicitly pass all of the variables or I get an undefined variable exception. For a component that passes no data to the layout, I still have to do this:

```PHP
public function render()
{
  return view('livewire.some-page')
            ->layoutData([
                'title' => null,
                'headline' => null,
                'sidebar' => null,
                'fullBleed' => false,
                'tabBarActions' => null,
                'breadcrumbLinks' => []
            ]);
}
```

First off, it's annoying to have to pass a bunch of null values. Secondly It's an issue because whenever a new layout variable  is introduced, all of the full page livewire components break, because they aren't set up to pass a value for that new piece of layout data.

Am I doing something wrong here? Is there a different way to set this up, where my livewire components work with default values from the layout component?

**Comment by calebporzio** [comment]
Based on what you're saying, I agree this should be fixed. Can you PR a fix with tests?

**Comment by cg0012** [comment]
Just structure your layout with optional parameters:

```
@if(isset($title))
   <div class"title stuff">
      {{ $title }}
   </div>
@endif
```

or create blade components and only include the ones you need on the views you need them:

```
<x-layout.title>
   {{ $title }}
</x-layout.title>
```

less html will be included in your views that don't use all your layout props. Could take a little time to clean up and implement but way better in the long run
